### DISCUSSION TITLE
UIKit Programmatic View Controller Presentation

### DISCUSSION THREAD

**Discussion by BastianKusserow** [pointfreeco/swift-composable-architecture#1222] [Q&A]
Hey,

I've already looked into the UIKit case studies but am somehow stuck how to get programmatic UIKit view controller presentation working.

Assume I've got the following state and actions:
```swift
struct MainState: Equatable {
    var next: DetailState?
}

struct DetailState: Equatable { }

enum MainAction {
    case next
    case detail(DetailAction)
}

enum DetailAction {
    case back
}
```
and the following reducers:
```swift
let mainReducer = Reducer< MainState, MainAction, Void> { state, action, _ in
    switch action {
    case .next:
        state.next = .init()
        return .none
    case .detail(.back):
        state.next = nil
        return .none
    case .detail:
        return .none
    }
}

let detailReducer = Reducer< DetailState, DetailAction, Void> { state, action, _ in
    switch action {
    case .back:
        return .none
    }
}

let appReducer = Reducer.combine(
    detailReducer
        .optional()
        .pullback(state: \.next, action: /MainAction.detail, environment: { () }),
    firstReducer
)
```

Now I'd like to present the detail screen if the user taps on a button in the main screen, which works just fine via scoping:
```swift
store?.scope(state: \.next, action: { .detail($0) })
            .ifLet { [weak self] localStore in
                let vc = DetailViewController(localStore)
                self?.present(vc, animated: true)
            } else: {
                self.dismiss(animated: true)
            }
            .store(in: &cancellables)
```
The problem now comes when trying to dismiss the detail view controller. Assuming one is using the standard UIKit sheet style, we have to be notified when the vc is dismissed when the user pulls down the sheet. 
Therefore I send a `.back` action in the detail view controllers `viewDidDisappear` and everything works fine.
But if I want to navigate programmatically, things get tricky. The `.back` action will be sent to the store when the user taps a button and the view controller gets dismissed properly. But after it is dismissed, the `.back` action gets sent again because `viewDidDisappear` is called and a runtime warning will be displayed because the DetailState is already nil.
Looking into the case studies I think I have to somehow handle that in the parent view controller but I am not sure how this should be done. Overriding `dismiss` in the parent and sending the `.detail(.back)` action there leads to some weird behaviours. 
I'd appreciate any help here!

**Comment by nayooti** [comment]
I think `viewDidDisappear` is not the right place to trigger the .back-action. Because it is also invoked if you move "forward" (pushing a new VC on the navigation stack. 
Play around with `UIAdaptivePresentationControllerDelegate`. The `presentationControllerDidAttemptToDismiss` and `presentationControllerDidDismiss` are triggered when pulling down the modal.
